text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+and a specified discriminant type:
+<https://doc.rust-lang.org/reference/type-layout.html#combining-primitive-representations-of-enums-with-fields-and-reprc>
+
+This layout consists of the discriminant, followed by a union of the
+representations of each `enum` variant that has fields.
+
+If an `enum` specifies `repr(crabi)` but does not specify a discriminant type,
+the `enum` is guaranteed to use the smallest discriminant type that holds the
+maximum discriminant value used by a variant in the `enum`.
+
+If the `enum` has no fields, or no fields with a non-zero size, crABI will
+represent the `enum` as only its discriminant.
+
+### Guaranteed niche optimization
+[niche]: #niche
+
+As a special case, if an `enum` using `repr(crabi)` has exactly two variants,
+one of which has no fields and the other of which has a single field, and the
+single field type has a specific type (defined below) with a "niche" value,
+then the `enum` is supported in crABI, and the `enum` representation uses
+"niche optimization" to have the same size as the field. For instance,
+`core::crabi::Option<T>` uses the niche optimization if `T` is one of the types
+supporting niche optimization.
+
+As another special case, if an `enum` using `repr(crabi)` has exactly two
+variants, both of which have a single field, and one of the two fields has type
+`()`, the `enum` is supported in crABI, and crABI applies the niche
+optimization to make the enum the same size as the field in the other variant.
+For instance, `core::crabi::Result<(), E>` uses the niche optimization if `E`
+is one of the types supporting niche optimization, and
+`core::crabi::Result<T, ()>` uses the niche optimization if `T` is one of the
+types supporting niche optimization.
+
+`core::crabi::Option` and `core::crabi::Result` are analogues to the standard
+`Option` and `Result` types, but defined with `repr(crabi)` and guaranteed to
+only use niche optimizations specified by crABI. These types support `?` and
+convenient bidirectional conversions to/from the Rust `Option` and `Result`
+types.
+
+As crABI must specify a precise stable layout, crABI's niche optimization only
+applies to the types explicitly listed in the crABI specification, even though
+the same type in the Rust ABI may be able to apply niche optimizations in more
+cases.
+
+Types that permit niche optimization (using `crabi::Option` as an example and
+using the niche value to represent `crabi::None`, both of which are abbreviated
+unqualified below for readability):
+- `Option<&T>`, `Option<&mut T>`, `Option<NonNull<T>>`, `Option<Box<T>>`, and
+  `Option` of any function pointer type are all passed using a null pointer to
+  represent `None`.
+- `Option<bool>` is passed using a single `u8`, where 0 is `Some(false)`, 1 is
+  `Some(true)`, and 2 is `None`.
+- `Option<char>` is passed using a single `u32`, where 0 through `0xD7FF` and
+  `0xE000` through `0x10FFFF` are possible `char` values, and `0x110000` is
```

Comment:
Is there any possibility that unicode could expand the range of  valid code points to include 0x110000? If so would `u32::max()` or one of the surragate pair numbers be better?

Reply:
> Is there any possibility that unicode could expand the range of valid code points to include 0x110000?

Rust assumes it will never happen. That is not the choice I would make, but it is the choice that Rust makes. The Unicode Consortium supposedly will never do this, and it will cause breakage if they do, but of course it is not literally impossible that they might reverse their commitment not to do this.

Reply:
With the way UTF-16 is designed, it's literally impossible to encode unicode scalar values `>= 0x110000`. UTF-8/32 have ways to encode those out-of-range values but unicode almost certainly never will use them because they want UTF-16 to keep working since it's used soo many places: Win32, Java, JavaScript, C#, VB.net, etc.

Reply:
`char` is a Unicode Scalar Value:
https://www.unicode.org/glossary/#unicode_scalar_value

Unicode scalar values outside the defined range are specifically forbidden by the Unicode Standard:
> Ill-formed: A Unicode code unit sequence that purports to be in a Unicode encoding form is called ill-formed if and only if it does not follow the specification of that Unicode encoding form.
> …
> Any code unit sequence that would correspond to a code point outside the defined range of Unicode scalar values would, for example, be ill-formed.

https://www.unicode.org/glossary/#ill_formed_code_unit_sequence

(So expanding the range would be a breaking change for a lot of existing Unicode software.)

Reply:
What is the reason 0x110000 is used for the niche rather than `u32::MAX`?